### PR TITLE
[sweet api][Kotlin] Improve conversion from native side to js

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Improve conversion between native and js objects.
From now, our API can handle more complex types as a promise result - like `List<Record>` or records with inner maps or other records.

# How

- Added a general way of doing conversion between native objects to `WritableMap` or to `WritableArray`.
- Made sure that the code is testable by adding `ContainerProvider`

# Test Plan

- unit tests ✅

### CODE REVIEW

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/JSTypeConverterHelper.kt [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    null, is Unit -> pushNull()
+    is ReadableArray -> pushArray(value)
+    is ReadableMap -> pushMap(value)
+    is String -> pushString(value)
+    is Int -> pushInt(value)
+    is Number -> pushDouble(value.toDouble())
+    is Boolean -> pushBoolean(value)
+    else -> throw IllegalArgumentException("Could not put ${value.javaClass} to 'WritableArray'")
+  }
+}
```
Comment (lukmccall) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/JSTypeConverterHelper.kt:148]:
```suggestion
}

```

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/JSTypeConverterHelper.kt [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+internal fun WritableMap.putGeneric(key: String, value: Any?) {
+  when (value) {
+    null, is Unit -> putNull(key)
+    is ReadableArray -> putArray(key, value)
+    is ReadableMap -> putMap(key, value)
+    is String -> putString(key, value)
+    is Int -> putInt(key, value)
+    is Number -> putDouble(key, value.toDouble())
+    is Boolean -> putBoolean(key, value)
+    else -> throw IllegalArgumentException("Could not put ${value.javaClass} to 'WritableMap'")
```
Comment (barthap) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/JSTypeConverterHelper.kt:133]:
```suggestion
    else -> throw IllegalArgumentException("Could not put '${value.javaClass}' to 'WritableMap'")
```

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/JSTypeConverterHelper.kt [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+internal fun WritableArray.putGeneric(value: Any?) {
+  when (value) {
+    null, is Unit -> pushNull()
+    is ReadableArray -> pushArray(value)
+    is ReadableMap -> pushMap(value)
+    is String -> pushString(value)
+    is Int -> pushInt(value)
+    is Number -> pushDouble(value.toDouble())
+    is Boolean -> pushBoolean(value)
+    else -> throw IllegalArgumentException("Could not put ${value.javaClass} to 'WritableArray'")
```
Comment (barthap) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/JSTypeConverterHelper.kt:146]:
```suggestion
    else -> throw IllegalArgumentException("Could not put '${value.javaClass}' to 'WritableArray'")
```

**Review by barthap — APPROVED** [review]
Nice one 👍 

Do you think we should get rid of the `EnumSerializerTest.kt` and merge its contents into `JSTypeConverterTest`?

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/JSTypeConverterHelper.kt (hunk 2 above)
Comment (tsapeta) [reply packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/JSTypeConverterHelper.kt:133]: or 👇 
```suggestion
    else -> throw IllegalArgumentException("Could not put '${value.javaClass}' to WritableMap")
```

The general rule is to quote dynamic/user's input strings.

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/JSTypeConverterHelper.kt (hunk 3 above)
Comment (tsapeta) [reply packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/JSTypeConverterHelper.kt:146]:
```suggestion
    else -> throw IllegalArgumentException("Could not put '${value.javaClass}' to WritableArray")
```

**Review by tsapeta — APPROVED** [review]

**Pull request merged**